builtin/logical/pki/path_manage_keys.go

```diff
@@ -0,0 +1,195 @@
+package pki
+
+import (
+	"context"
+	"strings"
+
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/certutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func pathGenerateKey(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "keys/generate/(internal|exported|kms)",
+
+		Fields: map[string]*framework.FieldSchema{
+			keyNameParam: {
+				Type:        framework.TypeString,
+				Description: "Optional name to be used for this key",
+			},
+			keyTypeParam: {
+				Type:        framework.TypeString,
+				Default:     "rsa",
+				Description: `Type of the secret key to generate`,
+			},
+			keyBitsParam: {
+				Type:        framework.TypeInt,
+				Default:     2048,
+				Description: `Type of the secret key to generate`,
+			},
+			"managed_key_name": {
+				Type: framework.TypeString,
+				Description: `The name of the managed key to use when the exported
+type is kms. When kms type is the key type, this field or managed_key_id
+is required. Ignored for other types.`,
+			},
+			"managed_key_id": {
+				Type: framework.TypeString,
+				Description: `The name of the managed key to use when the exported
+type is kms. When kms type is the key type, this field or managed_key_name
+is required. Ignored for other types.`,
+			},
+		},
+
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback:                    b.pathGenerateKeyHandler,
+				ForwardPerformanceStandby:   true,
+				ForwardPerformanceSecondary: true,
+			},
+		},
+
+		HelpSynopsis:    pathGenerateKeyHelpSyn,
+		HelpDescription: pathGenerateKeyHelpDesc,
+	}
+}
+
+const (
+	pathGenerateKeyHelpSyn  = `Generate a new private key used for signing.`
+	pathGenerateKeyHelpDesc = `This endpoint will generate a new key pair of the specified type (internal, exported, or kms).`
+)
+
+func (b *backend) pathGenerateKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// Since we're planning on updating issuers here, grab the lock so we've
+	// got a consistent view.
+	b.issuersLock.Lock()
+	defer b.issuersLock.Unlock()
+
+	keyName, err := getKeyName(ctx, req.Storage, data)
+	if err != nil { // Fail Immediately if Key Name is in Use, etc...
+		return nil, err
+	}
+
+	mkc := newManagedKeyContext(ctx, b, req.MountPoint)
+	exportPrivateKey := false
+	var keyBundle certutil.KeyBundle
+	var actualPrivateKeyType certutil.PrivateKeyType
+	switch {
+	case strings.HasSuffix(req.Path, "/exported"):
+		exportPrivateKey = true
+		fallthrough
+	case strings.HasSuffix(req.Path, "/internal"):
+		keyType := data.Get(keyTypeParam).(string)
+		keyBits := data.Get(keyBitsParam).(int)
```

Comment:
I'm adding extra input validation to keyType, keyBits

Reply:
In #15319.